### ISSUE TITLE
Costumer id document as custom field

### ISSUE THREAD

**Issue by Saiuke** [opensourcepos/opensourcepos#1983]
### Background information

__IMPORTANT: If you choose to ignore this issue report template, your issue will be closed as we cannot help without the requested information.__

Please make sure you tick (add an x between the square brackets with no spaces) the following check boxes:

- [X] Reporting an issue of an unmodified OSPOS installation
- [X] Checked [open and closed issues](https://github.com/opensourcepos/opensourcepos/issues?utf8=%E2%9C%93&q=is%3Aissue) and no similar issue was already reported (please make sure you searched!)
- [X] Read [README](https://github.com/opensourcepos/opensourcepos/blob/master/README.md), [WHATS_NEW](https://github.com/opensourcepos/opensourcepos/blob/master/WHATS_NEW.txt), [INSTALL.md](https://github.com/opensourcepos/opensourcepos/blob/master/INSTALL.md) and [UPGRADE](https://github.com/opensourcepos/opensourcepos/blob/master/UPGRADE.txt)
- [X] Read the [FAQ](https://github.com/opensourcepos/opensourcepos#faq) for any known install and/or upgrade gotchas (in specific PHP extensions installed)
- [X] Read the [wiki](https://github.com/opensourcepos/opensourcepos/wiki)
- [] Executed any database upgrade scripts if an upgrade pre 3.0.0 (e.g. database/2.4_to_3.0.sql)
- [X] Aware the installation code is in [bintray](https://bintray.com/jekkos/opensourcepos/opensourcepos/view/files?sort=updated&order=asc#files) (see README), and [GitHub master](https://github.com/opensourcepos/opensourcepos/tree/master) is for [developers only](https://github.com/opensourcepos/opensourcepos/wiki/Development-setup) and therefore not complete nor stable

### Installation information

- OSPOS version is: 3.2.0
- OSPOS git commit hash is: 4f5ad57
- PHP version is: 7.2
- MySQL or MariaDB version is: 5.7.14
- OS and version is: Windows 10
- WebServer is: Apache 2.4.23
- Selected language is: Spanish
- (If applicable) Docker installation: 
- (If applicable) Installation package for the LAMP/LEMP stack is: 

### New Feature

In many countries, is required to be displayed in the invoice or receipt the costumer's ID number, such as social security number, "cédula de identidad" or "CPF". Due to diferent types of documents in different countries, would be nice to have one or two custom fields in the client area. This could be done in a way that the user could customize the name of such fields and choose to print them or not.

Another fields, such as adress and telephone, should be optional to print too. In some business, where there is delivery, is necessary to have the client adress in the receipt.

**Comment by mckaygerhard** [comment]
umm yes , seems you are from Venezuela or Colombia

**Comment by daN4cat** [comment]
As explained in #1957 you can use the Customer Account Number (or Account#). See bottom of Customer input form.

**Comment by Saiuke** [comment]
Actually from Brazil, but I have some business in Colombia. So, I can use the account number as explained in #1957, but it is not the ideal way to do it. I mean, it should be a intentional feature. And, it does not appear in the conventional receipt.

**Comment by Saiuke** [comment]
I was thinking, not as a permanent solution, to change the name of the field adress2 to (in my specific case) NIT and add this to the invoice. How can I do this?

**Comment by mckaygerhard** [comment]
sorry but your case its not common, and i one of those that need it!

your case as my, its an ad-hoc need, so make a extra column in table and make extra fields inputs in the customer and in the invoice., not so dificult, for more info see the development index in the wiki

**Comment by daN4cat** [comment]
You could just add the option in the config to print the account number and print in the receipt the $company_info as in the invoice (that includes customer address, phone number and account number).

**Comment by Saiuke** [comment]
I have to add this custom field in a lot of places to make it work properly. As far as I know I would have to add the extra field in the database (costumers table I guess), in the new costumer scritp, in the edit costumer script, in the invoice, sales and so on... My point is, it should be way more simples with a simples custom field. That way anybody would be able to personalize the costumers data.

Anyway, I would require some guidance about how to do it, I'm not familiarized with MDV projects, I just know basic PHP and MySQL.

**Comment by mckaygerhard** [comment]
start to read the wiki, close the project as i pointed in previous comment.. later put more ask..

and please have advised that all the tools and developers need are for unix based not guindows..

**Comment by Saiuke** [comment]
Yeah, I'm going to migrate my stuff to Ubuntu or something alike. Anyway, thanks for your time and please consider marking my suggestion as a enhancement. It would help a lot.

**Comment by Saiuke** [comment]
I was able to modify the field name to NIT, as I needed, but I'm still no able to show it in the receipt. I'm working in the receipt file contained in application >> views >> sales >> receipt_default.php. I inserted this line after the line 39:
```
<div id="customer"><?php
            $customer_data['account_number'] = $account_number;
            echo $this->lang->line('customers_account_number').": ".$account_number; ?></div>
```
But the field (NIT in this case) ends up being empty when I print the receipt:

![screenshot-2018-5-12 ferreteria lopes camacho potenciado por ospos 3 2 0](https://user-images.githubusercontent.com/19313767/39961179-a5b2fa62-55f6-11e8-922b-b3586d230183.png)

And, of course, I added this information in the costumer form, as you can see below:

![screenshot-2018-5-12 ferreteria lopes camacho potenciado por ospos 3 2 0 1](https://user-images.githubusercontent.com/19313767/39961188-d7702192-55f6-11e8-9d76-89092d373bcd.png)

**Comment by mckaygerhard** [comment]
must be passed to the views with the data var, i recomended to read a little about codeigniter, its very easy, the data var its an array that handle allthe views variables

**Comment by mckaygerhard** [comment]
we need to classify the client too, as i remember in Colombia hay Juridico y natural, o gubernamental?

**Comment by mckaygerhard** [comment]
and @Saiuke dont forgit to pass the diff to put coding as contribution

**Comment by Saiuke** [comment]
I'll try to make something really usefull and functional and then make a commit, but it might take a while.
As I said, would be nice to have more than one custom field, this way the system could handle all sort of legal requirements in different countries. In my case, I don't think I need to classify my clients in differente categories. But, that's an idea. Normally it can be "Natural", "Jurídica" or "Governamental".

**Comment by Saiuke** [comment]
As son I get to modify the field and make it appear in the receipt I post how I made it here.

**Comment by mckaygerhard** [comment]
@Saiuke the variable must be parsed in the data array when the view are called.. all the views in codeigniter handle a special var that again, handle all the variables ... just add this variable to that array just before the $this->view(xxx, $array) are called

**Comment by mckaygerhard** [comment]
current customer info or customer from specific sale retreived?

in the first case depend of if the operator (employee) has already added a customer..  and must be extracted from the session sale object..

in the second case, retrieve the sale object from the sale lib using the spoecific sale id and then retrieve from there the customer part.. 

here right now exactly the code i do not have, but after you retireve the variable must then parse as you said:
$data['customer_account_number'] = $customer_info->account_number;
and thist just before the call of the view default_receipt.php , of course i dont know where are you calling that view..

**Comment by Saiuke** [comment]
Yeah... But, already got it. Thanks.
Just added this line after the line 39 in the receipt_default.php:

`<div id="customer"><?php echo $this->lang->line('customers_account_number').": ".$customer_account_number; ?></div>`

Turns out my first aproach was correct all the time. It was some sort of typping error or something.

**Comment by Saiuke** [comment]
Anyway, seems like we need a owner to the enhancement proposed. Thanks for the help so far.

**Comment by mckaygerhard** [comment]
good! i must made the same for Dominican Republic too.. i have a branch for that, i'll copy to you in some weeks when done, due i'm so busy right now!

**Comment by mckaygerhard** [comment]
umm but @Saiuke what other changes do you made already?

**Comment by Saiuke** [comment]
Well, I changed the translation files to what I needed, it was "account number" and I changed it to "NIT".
And added this line after the line 42 in the views/sales/receipt_default.php:
```
<?php
		if($customer_account_number!= '')
		{
?>
   <div id="customer"><?php echo $this->lang->line('customers_account_number').": ".$customer_account_number; ?></div>
                        
<?php
	         }
?>
```
Using this code you have the NIT (or whatever kind of document you want) in the receipt when the variable $customer_account_number isn't empty. The result you can see below:

![screenshot-2018-5-12 ferreteria lopes camacho potenciado por ospos 3 2 0 2](https://user-images.githubusercontent.com/19313767/39962631-5f1b6f3e-561a-11e8-8a07-b8e9f2cf217f.png)

**Comment by Saiuke** [comment]
I believe that adding the option to customize the name of the account number field in the client form and add this line to the receipt would do it.

**Comment by mckaygerhard** [comment]
but that only one field, we need two!

**Comment by SteveIreland** [comment]
I'm not clear on where this issue is heading.

I would also like to see maybe 5 custom fields added to the customer table similar to what we have now for items.  In my particular scenario for the RV service industry it would be nice to have a specific field to enter the critical information about vehicle identification number, model, etc

Doing a custom change for your own internal use is certainly fine (we all have them) but it does introduce ongoing upgrade challenges.

I think this should be a reasonably easy change since it would just involve mirroring the relevant code for item maintenance over to customer maintenance.

**Comment by Saiuke** [comment]
@SteveIreland. I agree. As I said, would be nice to have custom fields as a default feature, considering that the project is intented to all sorts os business and countries. I would like to know how to implement a final solution and share it with you, but I don't. For now.

**Comment by mckaygerhard** [comment]
i have working on that but in first shot will not be custom fields... i need this on a hurry so will be two "id" and "NIT¡" optional fields

**Comment by Saiuke** [comment]
Right, why don't you use the adress line 2 or some another "almost not used field"?

**Comment by SteveIreland** [comment]
If I can find an hour or two this week I'll submit a pull request to add four custom fields to the customer table.

**Comment by jekkos** [comment]
Once again #68. I picked up this story once again (progressed a bit on thursday) but still need to polish and test.

**Comment by SteveIreland** [comment]
Hmmm ... it was made very clear to me in #1193 that #68 was for items and not customers.

Of course, that was back in March of 2017 and things change.   :-)    Are you adapting it for customers too?

**Comment by mckaygerhard** [comment]
yeah @jekkos that was for items, and #1193 are for customers.. do you are working in customers custom fields or in items custom fields?

**Comment by Saiuke** [comment]
I'm quite sure that I would be able to develop it properly, does this feature fits the next release plan?

**Comment by jekkos** [comment]
@Saiuke we have a change already pending that would allow you to add this field in a generic way. #3161 is still open and planned for merge after we merge the item import functionality from @objecttothis .

So what you can do is check that PR, test it and see if it works similar to the item attributes. Those item attributes have a set of visibility definitions that toggle visibility for the values on a sales/receiving receipt.

I'd do this change in the same way as how the item attributes were implemented. Any feedback welcome in #3161
